lib/elixir/lib/integer.ex

```diff
@@ -238,67 +238,51 @@ defmodule Integer do
   def parse("", base) when base in 2..36,
     do: :error
 
-  def parse(binary, base) when is_binary(binary) and base in 2..36 do
-    parse_in_base(binary, base)
+  def parse(<<bin::binary>>, base) when base in 2..36 do
+    parse_sign(bin, base)
```

Comment:
Actually we can keep the single context here. Also ill move the parse_sign/2 logic into parse/2.